packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,295 @@
+import path from "node:path";
+import fg from "fast-glob";
+
+import type { ConfigRoute, DefineRouteFunction, RouteManifest } from "./routes";
+import { createRouteId, defineRoutes } from "./routes";
+import {
+  escapeEnd,
+  escapeStart,
+  isSegmentSeparator,
+  optionalEnd,
+  optionalStart,
+  paramPrefixChar,
+  routeModuleExts,
+} from "./routesConvention";
+
+export function flatRoutes(
+  appDirectory: string,
+  ignoredFilePatterns?: string[]
+): RouteManifest {
+  let extensions = routeModuleExts.join(",");
+
+  let routePaths = fg.sync(`*{${extensions}}`, {
+    absolute: true,
+    cwd: path.join(appDirectory, "routes"),
+    ignore: ignoredFilePatterns,
+  });
+
+  return flatRoutesUniversal(appDirectory, routePaths);
+}
+
+interface RouteInfo extends ConfigRoute {
+  name: string;
+  segments: string[];
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory - The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routeMap = new Map<string, RouteInfo>();
+  let nameMap = new Map<string, RouteInfo>();
+
+  for (let routePath of routePaths) {
+    let routeInfo = getRouteInfo(appDirectory, prefix, routePath);
+    routeMap.set(routeInfo.id, routeInfo);
+    nameMap.set(routeInfo.name, routeInfo);
+  }
+
+  // update parentIds for all routes
+  Array.from(routeMap.values()).forEach((routeInfo) => {
```

Comment:
You could use the native `for...of` loop for the `values()` iterator:

```suggestion
  for (let routeInfo of routeMap.values()) {
```

Kinda disappointing that iterators don't implement `forEach`...